File: aes/aes-256-gcm.ts
import { BinaryLike, createCipheriv, createDecipheriv, pbkdf2Sync, randomBytes } from 'crypto'

export class AES256GCM {
    public encrypt(text: string, masterkey: BinaryLike) {
        const iv = randomBytes(16)
        const salt = randomBytes(64)

        const key = pbkdf2Sync(masterkey, salt, 2145, 32, 'sha512')

        const cipher = createCipheriv('aes-256-gcm', key, iv)

        const encrypted = Buffer.concat([cipher.update(text, 'utf-8'), cipher.final()])

        const tag = cipher.getAuthTag()

        return Buffer.concat([salt, iv, tag, encrypted]).toString('base64')
    }

    public decrypt(encdata: string, masterkey: BinaryLike) {
        const bData = Buffer.from(encdata, 'base64')

        const salt = bData.slice(0, 64)
        const iv = bData.slice(64, 80)
        const tag = bData.slice(80, 96)
        const encrypted = bData.slice(96)

        const key = pbkdf2Sync(masterkey, salt, 2145, 32, 'sha512')

        const decipher = createDecipheriv('aes-256-gcm', key, iv)
        decipher.setAuthTag(tag)

        const decrypted = decipher.update(encrypted) + decipher.final('utf8')

        return decrypted
    }
}

File: py/aes_256_gcm.py
import base64
import hashlib
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes


class AES256GCM:
    def encrypt(self, text, masterkey):
        iv = get_random_bytes(16)
        salt = get_random_bytes(64)

        key = hashlib.pbkdf2_hmac(
            "sha512", masterkey.encode("utf-8"), salt, 2145, dklen=32
        )

        cipher = AES.new(key, AES.MODE_GCM, iv)

        encrypted = cipher.encrypt(text.encode("utf-8"))

        tag = cipher.digest()

        encrypted_data = salt + iv + tag + encrypted

        return base64.b64encode(encrypted_data).decode("utf-8")

    def decrypt(self, encdata, masterkey):
        bData = base64.b64decode(encdata)

        salt = bData[:64]
        iv = bData[64:80]
        tag = bData[80:96]
        text = bData[96:]

        key = hashlib.pbkdf2_hmac(
            "sha512", masterkey.encode("utf-8"), salt, 2145, dklen=32
        )

        cipher = AES.new(key, AES.MODE_GCM, iv)
        cipher.update(tag)

        decrypted = cipher.decrypt(text).decode("utf-8")

        return decrypted
